File: app/src/main/java/com/example/walmartpathfinder/models/Section.kt
package com.example.walmartpathfinder.models

abstract class Section(
    private var name: String,
    private var parent: Section?,
) {
    abstract fun sortItemList(unorderedList: MutableList<Item>): Pair<MutableList<Item>,MutableList<Item>>

    /* Check to see if product is the Descendant of the section and subsection */
    fun isDescendentOf(section: Section): Boolean {
        var cur = this
        while (cur.parent != null) {
            cur = cur.parent!!
            if (cur == section) {
                return true
            }
        }
        return false
    }

    fun setParent(parent : Section) {
        this.parent = parent
    }

    override fun toString(): String {
        return if (this.parent == null) (this.name) else (this.parent!!.name + "-" + this.name)
    }

}

File: src/models/Section.ts
import { Item } from "./Item";

abstract class Section{
    name: string | undefined;
    parent: Section | undefined;
    
    isDescendantOf(section: Section): Boolean{
        let cur = this as Section;
        while(cur.parent != null){
            cur = cur.parent;
            if (cur == section){
                return true;
            }
        }
        return false;
    } 

    setParent(parent: Section): void{
        this.parent = parent;
    }

    abstract sortItemList(unorderedList: Array<Item>): [Array<Item>, Array<Item>]
}


export {Section};